### PR TITLE
Model Documentation

### LINKED ISSUE #259: Modals for objects descriptions/documentation
### Environment
* Python version:  3.7
* Peering Manager version:  master

### Proposed Functionality

Add a new help button to every create/update forms for each model popping up a modal explaining what the model represents and what are the purposes of every fields. The documentation for each model must also be available in the doc website.

### Use Case

This feature is for helping people in understanding what are the relationships between models and how they interact between each other.

### External Dependencies

None.

### PR DESCRIPTION
**Pull request by respawner**
### Fixes:

Closes #259

Write documentation for models which are publicly exposed.

- [x] Write documentation for models
- [x] Add button in create/update forms to display the doc of the current model
- [x] Expose documentation on the public doc website

### CODE REVIEW

File: docs/models/utils/tag.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Tags
+
+Tags are text labels which can be applied to objects. Each tag has a label and
+a URL-friendly slug. The slug is generated automatically and makes tags easier
+to work with as URL parameters. Objects can be filtered by tags.
+
+For each tag that you create, the following properties are to be provided (some
+are however optional):
+
+  * `Name`: a human readable name attached to a tag.
```
Comment (LindsayHill) [inline docs/models/utils/tag.md:10]: s/human readable/human-readable/

File: docs/models/utils/tag.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  * `Slug`: a unique configuration and URL friendly name; most of the time it
```
Comment (LindsayHill) [inline docs/models/utils/tag.md:11]: Suggest "...and URL-friendly name. Usually automatically generated from the tag's name"

File: docs/models/utils/tag.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    is automatically generated from the tag's name.
+  * `Color`: a RGB color code that will be used when displaying the tag in the
+    user interface.
+  * `Comments`: some text that can be formatted in Markdown to explain what the
```
Comment (LindsayHill) [inline docs/models/utils/tag.md:15]: Suggest:
"text to explain what the tag is for. Can use Markdown formatting."

File: docs/models/users/token.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+A token is a unique identifier that identifies a user to the API. Each user in
+Peering Manager can have one or more tokens to authenticate to the API. To
+create a token, navigate to the API tokens page at `/user/api-tokens/`.
+
+Each token contains a 160-bit key represented as 40 hexadecimal characters.
+When creating a token, leaving the key field blank will automatically generate
+a random key.
+
+By default, a token can be used for all operations available via the API. It
+can be restricted only to read operations by deselecting the "write enable"
```
Comment (LindsayHill) [inline docs/models/users/token.md:12]: "...can be restricted to read-only operations by deselecting..."

File: docs/models/users/token.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+options.
+
+A token can also be programmed to expire at a specific time. This can be
```
Comment (LindsayHill) [inline docs/models/users/token.md:15]: "A token can be given an expiry time. This is useful if an external client only needs temporary access"

File: docs/models/peering/template.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Template
+
+A template is a text that can be parsed using a context processor that will
```
Comment (LindsayHill) [inline docs/models/peering/template.md:3]: Something missing here "A template is a text <something> that can be parsed...". Not sure what the best word is.

Maybe phrase as this:

"Templates are used for rendering device configurations and emails. The template is parsed using a context processor to complete variable references and execute control statements."

File: docs/models/peering/template.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+give a rendered version of the text by completing variable references and
+executing control statements.
+
+Inside Peering Manager, templates use the
+[Jinja2](https://palletsprojects.com/p/jinja/) syntax which allows for complex
+logics building. Depending if the template is for a configuration or an e-mail
```
Comment (LindsayHill) [inline docs/models/peering/template.md:9]: s/logics/logic/

File: docs/models/peering/template.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+different variables will be exposed to the Jinja2 processor.
```
Comment (LindsayHill) [inline docs/models/peering/template.md:10]: For second sentence, maybe try:
"Different variables are exposed to the Jinja2 processor for email and configuration templates"

File: docs/models/peering/template.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+Examples of templates are provided in the Peering Manager's
+[documentation](https://peering-manager.readthedocs.io/en/latest/templates/).
+
+For each template that you create, the following properties are to be provided
```
Comment (LindsayHill) [inline docs/models/peering/template.md:15]: "...the following properties can be configured (n.b. some are optional):"

File: docs/models/peering/template.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+(some are however optional):
+
+  * `Name`: a human readable name attached to a template.
```
Comment (LindsayHill) [inline docs/models/peering/template.md:18]: /s/human readable/human-readable/

File: docs/models/peering/template.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  * `Slug`: a unique configuration and URL friendly name; most of the time it
+    is automatically generated from the template's name.
+  * `Type`: a value that tells if the template is for configurations or
```
Comment (LindsayHill) [inline docs/models/peering/template.md:21]: "identifies if the template is for device configuration or email"

File: docs/models/peering/template.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    e-mails.
+  * `Template`: an actual text formatted according to the Jinja2 syntax.
```
Comment (LindsayHill) [inline docs/models/peering/template.md:23]: "The template itself, formatted using Jinja2 syntax"

File: docs/models/peering/template.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  * `Comments`: some text that can be formatted in Markdown to explain what the
```
Comment (LindsayHill) [inline docs/models/peering/template.md:24]: text to explain what the template is for. Can use Markdown formatting."

File: docs/models/peering/routingpolicy.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
+# Routing Policy
+
+Routing policies are a set of instructions to apply when advertising or
+receiving prefixes from another autonomous system. They are meant to filter
+routes to import or export. Routing policies play an important role to secure
+BGP by rejecting routes that does not belong in the Internet. Policies can in
+most cases be chained to achieve a complex and effective filtering process
+while keeping policy as simple as possible.
```
Comment (LindsayHill) [inline docs/models/peering/routingpolicy.md:8]: "Routing policies play an important role in securing BGP by rejecting invalid routes. Policies can be be chained to achieve complex and effective filtering, while keeping policies as simple as possible"

File: docs/models/peering/routingpolicy.md [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+## In Peering Manager
+
+Inside Peering Manager, you create routing policies by giving each of them a
+name. Policies must already exist on the router as their logics cannot be
```
Comment (LindsayHill) [inline docs/models/peering/routingpolicy.md:13]: s/logics/logic/

File: docs/models/peering/autonomoussystem.md [hunk 16]
```diff
@@ -0,0 +1,41 @@
+# Autonomous System
+
+An autonomous system, also known as AS, is a collection of connected Internet
+Protocol resources managed by network operators. It represents a single
+administrative entity which has clearly identified routing policies.
+
+An AS is defined mainly by its number, called the ASN. It is a 32-bit integer
+to each autonomous system. Some numbers are reserved for private use and must
```
Comment (LindsayHill) [inline docs/models/peering/autonomoussystem.md:8]: Probably just "It is a 32-bit integer (0 to 4294967295)"

File: docs/models/peering/autonomoussystem.md [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+## In Peering Manager
+
+Inside Peering Manager, you create autonomous systems that you want to peer
+with. Your own ASN number must be provided as a setting called `MY_ASN`. For
+each AS that you create, the following properties are to be provided (some are
+however optional):
+
+  * `ASN`: a unique autonomous system number of the entity.
+  * `Name`: a human readable name attached to an AS.
+  * `IRR AS-SET`: one or more words giving other autonomous systems that can be
```
Comment (LindsayHill) [inline docs/models/peering/autonomoussystem.md:20]: Maybe: "The set of other ASNs that can be found behind this AS" ?

File: docs/models/peering/autonomoussystem.md [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    found, for a BGP speaker, behind this AS.
+  * `IPv6 Max Prefixes`: a zero or positive integer which defines the maximum
+    number of IPv6 prefixes to receive from the AS.
+  * `IPv4 Max Prefixes`: a zero or positive integer which defines the maximum
+    number of IPv4 prefixes to receive from the AS.
+  * `Contact Name`: a name for a contact in charge of BGP/peering operations.
```
Comment (LindsayHill) [inline docs/models/peering/autonomoussystem.md:26]: "Contact person/team in charge of BGP/peering operations"

File: docs/models/peering/bgpgroup.md [hunk 19]
```diff
@@ -0,0 +1,37 @@
+# BGP Group
+
+A group of BGP sessions that share, more or less, the same purpose. It is a
+common practice to group BGP sessions together. For instance, one may want to
+make a group composed of transit provider BGP sessions only. A group can be
+defined once but it is not tight to a router. This means that one group can be
+configured on several routers depending on to which router the BGP sessions are
```
Comment (LindsayHill) [inline docs/models/peering/bgpgroup.md:7]: "...depending on which router the BGP sessions...."

File: docs/models/peering/bgpgroup.md [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+attached.
+
+## In Peering Manager
+
+Inside Peering Manager, you create a BGP group to aggregate multiple direct BGP
+sessions that you want to manage in a same way. A common use case would be to
```
Comment (LindsayHill) [inline docs/models/peering/bgpgroup.md:13]: s/manage in a same way/manage in the same way/

File: docs/models/peering/bgpgroup.md [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+create one BGP group for transit providers, one for customers and one for
+private peering sessions. You can create as many group as you want depending on
```
Comment (LindsayHill) [inline docs/models/peering/bgpgroup.md:15]: "many groups"

**Comment by respawner** [comment]
Suggested changes have been merged. Thanks for the feedback. If you want to change anything more, feel free to do so :)

**Comment by LindsayHill** [comment]
Thanks, I'll take another look in a couple of days. Tired eyes from looking at too many spreadsheets today 🤪

**Pull request merged**